File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int minDays(vector<int>& nums, int m, int k) {
        int n = nums.size();
        int left = 1;
        int right = 1e9;
        
        if((long long)m*k > n) return -1;
        
        while(left <= right) {
            int mid = (left + right) / 2;
            int flower = 0, bouquet = 0;
            
            for (int i = 0; i < n; i++) {
                if(nums[i] > mid) {
                    flower = 0;
                } else if (++flower >= k) {
                    bouquet++;
                    flower = 0;
                }
            
            }
            if(bouquet < m){
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return left;
    }
};

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.js
/**
 * @param {number[]} nums
 * @param {number} m
 * @param {number} k
 * @return {number}
 */
var minDays = function(nums, m, k) {
    let n = nums.length;

    let low = 1;
    let high = Math.max(...nums);
    let ans = -1;
    
    while(low <= high){
        let mid = Math.floor(low + (high - low) / 2);
        let flower = 0;
        let bouquet = 0;
        for(var i = 0; i < n; i++){
            if(nums[i] <= mid){
                flower++;
                if(flower == k){
                    bouquet++;
                    flower = 0;
                }
            }
            else flower = 0;
        }
        if(bouquet >= m) {
            ans = mid;
            high = mid - 1;
        } else {
            low = mid +1
        }
    }
    return ans;
};